packages/forms/src/patterns/attachment/response.ts

```diff
@@ -0,0 +1,116 @@
+
+  // Define the schema for a single file in the FileList
+  const fileSchema = z
+    .instanceof(File)
+    .refine((file: File) => data.allowedFileTypes.includes(file.type), {
+      message: `Invalid file type`,
+    })
+    .refine((file: File) => file.size <= maxFileSizeBytes, {
+      message: `Each file must be smaller than ${data.maxFileSizeMB} MB`,
+    });
+
+  // Define the schema for the FileList as a whole
+  return z
+    .custom<FileList>(
+      (fileList: FileList) => {
+        if (!(fileList instanceof FileList)) {
+          return false;
+        } else if (data.required && fileList.length === 0) {
+          return false;
+        } else if (fileList.length > data.maxAttachments) {
+          return false;
+        }
+        return true;
+      },
+      {
+        message:
+          data.maxAttachments === 1
+            ? `You must provide 1 file`
+            : `You must provide between 1 and ${data.maxAttachments} file(s).`,
+      }
+    )
+    .superRefine((fileList: FileList, ctx: z.RefinementCtx) => {
+      Array.from(fileList).forEach(file => {
+        const result = fileSchema.safeParse(file);
+        if (!result.success) {
+          result.error.errors.forEach((err: z.ZodIssue) => {
+            ctx.addIssue({
+              code: z.ZodIssueCode.custom,
+              message: err.message,
+            });
+          });
+        }
+      });
+    });
+};
+
+type IsBrowser = typeof window extends typeof globalThis ? true : false;
+
+// Conditionally type this pattern based on environment
+export type AttachmentPatternOutput = IsBrowser extends true
+  ? z.infer<ReturnType<typeof createFileListSchema>>
+  : z.infer<ReturnType<typeof createPostSchema>>;
+
+const createPostSchema = (data: AttachmentPattern['data']) => {
+  const maxFileSizeBytes = convertMBtoBytes(data.maxFileSizeMB);
+
+  return z
+    .any()
+    .refine(
+      (items: File[]) => {
+        return !(data.required && items.length === 0);
+      },
+      {
+        message: 'This field is required',
+      }
+    )
+    .refine(
+      (items: File[]) => {
+        return items.every(item => data.allowedFileTypes.includes(item.type));
+      },
+      {
+        message: `Invalid file type`,
+      }
+    )
+    .refine(
+      (items: File[]) => {
+        return items.length <= data.maxAttachments;
+      },
+      {
+        message: `The maximum number of attachments is ${data.maxAttachments}`,
+      }
+    )
+    .refine(
+      (items: File[]) => {
+        return items.every(item => item.size <= maxFileSizeBytes);
+      },
+      {
+        message: `The maximum allowable size per file is ${data.maxFileSizeMB}MB`,
+      }
+    );
+};
+
+export const parseUserInput: ParseUserInput<
+  AttachmentPattern,
+  AttachmentPatternOutput
+> = (pattern, obj) => {
+  return safeZodParseToFormError(
+    isBrowser
+      ? createFileListSchema(pattern['data'])
+      : createPostSchema(pattern['data']),
+    obj
+  );
```

Comment:
I'm glad you tracked down a solution here. I'm wondering if there may be some unintended consequences of this type being different in the different scenarios. One way to avoid might be to have the schema first coerce a `FileList` into `File[]`?

Reply:
I removed the branching logic from the validation. It now works on client and server without having separate conditions for each.